pkg/controller/privatelink/actuator/gcpactuator/gcplinkactuator_test.go

```diff
@@ -0,0 +1,2450 @@
+	}
+
+	mockForwardingRule = &compute.ForwardingRule{
+		Name:       "testforwardingrule",
+		IPAddress:  mockAddress.Address,
+		Region:     testRegion,
+		Target:     "testtarget",
+		Subnetwork: mockSubnet.Name,
+		SelfLink:   "testforwardingruleselflink",
+	}
+
+	mockServiceAttachment = &compute.ServiceAttachment{
+		Name:     "testserviceattachment",
+		Region:   testRegion,
+		SelfLink: "testserviceattachmentselflink",
+	}
+)
+
+func newTestGCPLinkActuator(t *testing.T, config *hivev1.GCPPrivateServiceConnectConfig, cd *hivev1.ClusterDeployment, existing []runtime.Object, configureGCPClient func(*mockclient.MockClient)) (*GCPLinkActuator, error) {
+	fakeClient := client.Client(testfake.NewFakeClientBuilder().
+		WithRuntimeObjects(cd).
+		WithRuntimeObjects(existing...).
+		Build())
+
+	mockedGCPClient := mockclient.NewMockClient(gomock.NewController(t))
+	if configureGCPClient != nil {
+		configureGCPClient(mockedGCPClient)
+	}
+
+	return &GCPLinkActuator{
+		client:         &fakeClient,
+		config:         config,
+		gcpClientHub:   mockedGCPClient,
+		gcpClientSpoke: mockedGCPClient,
+	}, nil
+}
+
+func newGoogleApiError(code int, reason string, message string) error {
+	err := errors.New(message)
+	herr := &googleapi.Error{
+		Code:    code,
+		Message: message,
+		Errors: []googleapi.ErrorItem{{
+			Reason:  reason,
+			Message: message,
+		}},
+	}
+	errors.As(err, &herr)
+	return herr
+}
+
+func Test_Cleanup(t *testing.T) {
+	cdBuilder := testcd.FullBuilder(testNameSpace, testClusterDeployment, scheme.GetScheme())
+
+	cases := []struct {
+		name string
+		cd   *hivev1.ClusterDeployment
+
+		gcpClientConfig func(*mockclient.MockClient)
+
+		expectError  string
+		expectStatus *hivev1gcp.PrivateServiceConnectStatus
+	}{{ // There should be an error on cleanupEndpoint failure
+		name: "cleanupEndpoint failure",
+		cd: cdBuilder.Build(
+			testcd.WithGCPPlatform(&hivev1gcp.Platform{}),
+			testcd.WithGCPPlatformStatus(&hivev1gcp.PlatformStatus{
+				PrivateServiceConnect: &hivev1gcp.PrivateServiceConnectStatus{Endpoint: "testEndpoint"},
+			}),
+			testcd.WithClusterMetadata(&hivev1.ClusterMetadata{InfraID: testInfraID}),
+		),
+		gcpClientConfig: func(m *mockclient.MockClient) {
+			m.EXPECT().DeleteForwardingRule(gomock.Any(), gomock.Any()).Return(newGoogleApiError(http.StatusUnauthorized, "AccessDenied", "not authorized to DeleteForwardingRule"))
+		},
+		expectError:  "error cleaning up endpoint: error deleting the Endpoint: googleapi: Error 401: not authorized to DeleteForwardingRule, AccessDenied",
+		expectStatus: &hivev1gcp.PrivateServiceConnectStatus{Endpoint: "testEndpoint"},
+	}, { // There should be an error on cleanupEndpointAddress failure
+		name: "cleanupEndpointAddress failure",
+		cd: cdBuilder.Build(
+			testcd.WithGCPPlatform(&hivev1gcp.Platform{}),
+			testcd.WithGCPPlatformStatus(&hivev1gcp.PlatformStatus{
+				PrivateServiceConnect: &hivev1gcp.PrivateServiceConnectStatus{EndpointAddress: "testEndpointAddress"},
```

Comment:
For this and subsequent tests, can we have the status initially contain all the _other_ values? Then the `expectStatus` containing only the "last" (failed) one proves that we successfully scrubbed the rest.

That said, we may want to consider making all the status updates locally and only pushing the `Status().Update()` once, at the end. Idempotence again. Noting in HIVE-2687.

Reply:
These cases are only ensuring the error state. The "success" cast tests successfully scrubbing them all. I don't see a benefit of turning this into a complex matrix.

We could update this to be more idempotent. However, I argue that under idempotent design, the status.update() should explicitly be done on each item and the function should return immediately with a requeue requested. With this design, there is no situation where "the rest" would ever be scrubbed. Even so, this is a task for HIVE-2687.

Reply:
> These cases are only ensuring the error state.

The other point of view is that the code has a behavior, and we should be validating that behavior such that, if it changes, we are forced to recognize that it has changed.

Is it actually important to the design whether the fields and conditions are lockstepped or bundled? _Probably_ not... but maybe.

But yah, we can discuss later.